### ISSUE TITLE
PointFeatureCollection broken by updating file

### ISSUE THREAD

**Issue by dopplershift** [Unidata/thredds#472]
On our systems, the METAR data are written into daily files. After requesting data from today's file, subsequent requests get an error once the file is updated with more data:
```
Illegal Range for dimension 0: last requested 106738 > max 98823
```

This can be demostrated using Siphon:
```python
from datetime import datetime
from siphon.catalog import TDSCatalog
from siphon.ncss import NCSS

cat = TDSCatalog('http://atm.ucar.edu/thredds/catalog/nws/metar/ncdecoded/catalog.xml')
ds = list(cat.datasets.values())[0]
ncss = NCSS(ds.access_urls['NetcdfSubset'])
print(ncss.variables)

query = ncss.query()
query.lonlat_box(east=-80, west=-90, south=25, north=35).accept('netcdf4')
query.variables('air_temperature', 'dew_point_temperature').time(datetime(2016,3,8,12))
nc = ncss.get_data(query)
```

My guess is that something is caching part of the file, but the request for the current length of the "recNum" unlimited dimension goes through to the actual file. We probably never encountered this before since collections were never properly updating (or at least we couldn't query them).

**Comment by dopplershift** [comment]
It seems to fix itself on the hour(?), which is odd, since the `FeatureCollection` tag tells it to check every 15 minutes thredds/atm.

**Comment by dopplershift** [comment]
So it looks like problematic sequence (through **so many** nested calls) is:

1. Load netcdf file header information for feature collection--this includes variables sizes
2. File on disk updates
3. Get request
4. Get station information, which necessitates reading information from disk
5. Station information includes the offsets for this linked-list-in-NetCDF.
6. Offsets from data on disk are greater than our (cached) size of variables
7. ~~Profit!~~ Exception

**Comment by dopplershift** [comment]
Options that I can think of:
- Abandon any support for this linked-list thingy
- Somehow make updating the collection pull in the station information. This would end up joining some code that's pretty far apart as far as I can tell
- Somehow make sure the NetcdfFile instances for the collection are not cached.
- ????
We're in this unique hell because not all of the header/metadata information is read in one "atomic" operation and we're simultaneously having the netcdf files, which are not message-based, updated on disk.

@JohnLCaron any thoughts?

**Comment by dopplershift** [comment]
Ok, a few more steps deeper, the problem comes down to `getStationHelper()`; while a few levels up I can find cached data with offsets that aren't updated (and not out of bounds), these don't end up going to the `StationHelper` instance, which triggers a new read from the file. Not sure how to make this cache or something...

**Comment by JohnLCaron** [comment]
Where does the netcdf file get written / updated? Is this Rob legacy? This is the backwards linked list thing?

**Comment by dopplershift** [comment]
It's written by the Robb script for METAR. The bigger problem is updating netcdf files + linked list stuff.

**Comment by JohnLCaron** [comment]
Probably easy enough to fix. ill look at it tommorrow i hope.

BTW,

http://thredds-dev.unidata.ucar.edu/thredds/fileServer/nws/metar/ncdecoded/files/Surface_METAR_20160207_0000.nc

is failing, but

http://thredds.ucar.edu/thredds/fileServer/nws/metar/ncdecoded/files/Surface_METAR_20160207_0000.nc

working

**Comment by JohnLCaron** [comment]
But really we should get rid of this legacy stuff. 
And replace it with alphanumerics, I mean BUFR, I mean something cool.

**Comment by dopplershift** [comment]
Replacing with BUFR would imply that we actually have a feed of US surface data in BUFR...which isn't coming across NOAAPORT that I've found so far. I'm still digging there.

We will be working to make extended model NetCDF4 files for CF-2.0. Either way, I will **kill** this legacy linked list thingy.

**Comment by JohnLCaron** [comment]
> Load netcdf file header information for feature collection--this includes variables sizes
> Station information includes the offsets for this linked-list-in-NetCDF

what class has variable size ?

what class has linked-list offset?

**Comment by dopplershift** [comment]
I'm not sure what you mean by class holding these things. The fundamental problem is that the `StationHelper` class ends up reading the `lastChild` variable from disk itself; the values from this variable are used to index into a `Structure` created for the PointFeatureCollection. Unfortunately, these indexes are bigger than the limits of the Structure.

Easiest way to reproduce. Make a point feature collection pointing at this file. Set a breakpoint at Section.java:900 (`checkInRange`). Boom.
[Surface_METAR_20160308_0000.nc.gz](https://github.com/Unidata/thredds/files/165480/Surface_METAR_20160308_0000.nc.gz)

**Comment by JohnLCaron** [comment]
> We will be working to make extended model NetCDF4 files for CF-2.0.

I guess the questions are:
1. is Unidata going to redo the TDS point data infrastructure? 
2. Who and when?
3. Use ERDAP (?!)
4. Meanwhile we need some bandaids. General or just for our own data?

**Comment by JohnLCaron** [comment]
>   Make a point feature collection pointing at this file. Set a breakpoint at Section.java:900 (checkInRange). 

do you have to update the file to see the problem?

if you have a stacktrace, can you send?

**Comment by dopplershift** [comment]
You do have to update, sorry. The file above is the file with the additional data. I've attached the first data.

[Surface_METAR_20160308_0000.nc-first.gz](https://github.com/Unidata/thredds/files/165489/Surface_METAR_20160308_0000.nc-first.gz)

**Comment by dopplershift** [comment]
Here's the stacktrace (snipping out Spring stuff):
```
2016-03-08T17:27:49.987 -0700 [   4432406][      14] ERROR - ucar.nc2.ft.point.StructureDataIteratorLinked - StructureDataLinkedIterator.nextStructureData recno=125574
ucar.ma2.InvalidRangeException: Illegal Range for dimension 0: last requested 125574 > max 111349
	at ucar.ma2.Section.fill(Section.java:179) ~[edu.ucar-cdm-4.6.5-SNAPSHOT.jar:?]
	at ucar.nc2.Variable.read(Variable.java:709) ~[edu.ucar-cdm-4.6.5-SNAPSHOT.jar:?]
	at ucar.nc2.Structure.readStructure(Structure.java:357) ~[edu.ucar-cdm-4.6.5-SNAPSHOT.jar:?]
	at ucar.nc2.ft.point.StructureDataIteratorLinked.next(StructureDataIteratorLinked.java:70) [edu.ucar-cdm-4.6.5-SNAPSHOT.jar:?]
	at ucar.nc2.ft.point.PointIteratorFromStructureData.nextStructureData(PointIteratorFromStructureData.java:109) [edu.ucar-cdm-4.6.5-SNAPSHOT.jar:?]
	at ucar.nc2.ft.point.PointIteratorFromStructureData.hasNext(PointIteratorFromStructureData.java:74) [edu.ucar-cdm-4.6.5-SNAPSHOT.jar:?]
	at ucar.nc2.ft.point.collection.CompositeStationCollection$CompositeStationFeature$CompositeStationFeatureIterator.hasNext(CompositeStationCollection.java:338) [edu.ucar-cdm-4.6.5-SNAPSHOT.jar:?]
	at ucar.nc2.ft.point.PointCollectionImpl.hasNext(PointCollectionImpl.java:128) [edu.ucar-cdm-4.6.5-SNAPSHOT.jar:?]
	at thredds.server.ncss.view.dsg.station.AbstractStationSubsetWriter.writeStationTimeSeriesFeature(AbstractStationSubsetWriter.java:96) [classes/:4.6.5-SNAPSHOT]
	at thredds.server.ncss.view.dsg.station.AbstractStationSubsetWriter.write(AbstractStationSubsetWriter.java:83) [classes/:4.6.5-SNAPSHOT]
	at thredds.server.ncss.view.dsg.AbstractDsgSubsetWriter.respond(AbstractDsgSubsetWriter.java:40) [classes/:4.6.5-SNAPSHOT]
	at thredds.server.ncss.controller.NcssController.handleRequestDsg(NcssController.java:232) [classes/:4.6.5-SNAPSHOT]
	at thredds.server.ncss.controller.NcssController.handleRequest(NcssController.java:147) [classes/:4.6.5-SNAPSHOT]
```

**Comment by JohnLCaron** [comment]
> We probably never encountered this before since collections were never properly updating (or at least we couldn't query them).

so we made a change so that now collections are updating, when they werent before? is there a commit for that change?

**Comment by dopplershift** [comment]
@lesserwhirls fixed the validation of time ranges in `NcssParamBean`. See 5cea2afb82daa9166a4e235eb46061338ee383d2.

**Comment by dopplershift** [comment]
It also worked before because we weren't properly caching the file, so everything went to disk I think. This is based on the fact that if I remove @cwardgar 's fixed for caching in 182c1ae (remove the `lastModified != 0`), everything is hunky dory.

**Comment by JohnLCaron** [comment]
@cwardgar can you explain how you understand this fix from 182c1ae ?

"a bug that caused FileCache to remove entries for closed datasets"

  long lastModified = want.ncfile.getLastModified();  // Will be 0 if ncfile is closed.
  boolean changed = lastModified != 0 && lastModified != want.lastModified;

1) i think comment should be

 // Will be 0 if raf is closed.

2) im my thinking, if the raf is closed, we need to open it and see if its changed. what may be happening is that because its closed, we assume not changed....

so, rehydrate then check if changed.

**Comment by cwardgar** [comment]
You can see an example of that fix working in [ReacquireClosedDatasetSpec](https://github.com/Unidata/thredds/blob/5.0.0/cdm/src/test/groovy/ucar/nc2/util/cache/ReacquireClosedDatasetSpec.groovy). Comment-out the `lastModified != 0 && ` bit in `FileCache` and the test fails.

I agree with your logic; we should be examining the raf and calling `raf.getLastModified()`. And indeed, that's what we're doing: follow the function calls beginning with `NetcdfFile.getLastModified()`.

However, I'm concerned about [AbstractIOServiceProvider.getLastModified()](https://github.com/Unidata/thredds/blob/5.0.0/cdm/src/main/java/ucar/nc2/iosp/AbstractIOServiceProvider.java#L158). It calls `raf.getLastModified()`, which is what we want, but only if `raf != null`. Unfortunately, we're assigning `null` to `raf` in the `close()` and `release()` methods. That seems like a problem. Any reason not to leave that reference intact?

**Comment by cwardgar** [comment]
Alternatively, we could change the implementation of `AbstractIOServiceProvider.getLastModified()` to:
```java
  @Override
  public long getLastModified() {
    File file = new File(location);
    return file.lastModified();
  }
```
which would make it match `RandomAccessFile.getLastModified()`.

**Comment by cwardgar** [comment]
I'm now convinced that I need to take a close look at all `getLastModified()` implementations and make sure they're doing the right thing. The `lastModified != 0 && ` bit in `FileCache` may not be necessary.

I'll work on that today.

**Comment by JohnLCaron** [comment]
Im looking at 4.6 FileCache, starting at line 349


if lastModified == 0, i dont see where raf is reacquired and then tested if
it changed....

**Comment by cwardgar** [comment]
`RandomAccessFile.getLastModified()` is sufficient to detect changes; no need to rehydrate first. We just have to make sure it's being called everywhere it should.

**Comment by JohnLCaron** [comment]
> Unfortunately, we're assigning null to raf in the close() and release()
 methods

once raf is closed, one should not use the reference anymore i think, so
setting to null seems right.

your idea of checking the file independent of raf might be sound, however.
havent follwed it through until i understand these other parts.

**Comment by JohnLCaron** [comment]
i think you are saying get rid of the return 0 possibility in FileCacheable,
by checking last modified independent of status of raf.
i think that could work.

**Comment by deeplycloudy** [comment]
Once the above commit is deployed to your test/dev server, please let me know and I'll be happy to try a remote request to help verify this fix.

**Comment by dopplershift** [comment]
@deeplycloudy I've put a new snapshot on http://atm.ucar.edu/thredds/. Kick the tires and let me know if you still see the problem.

**Comment by dopplershift** [comment]
Sweet. Closing (Fixed by #479)
